### PR TITLE
Skip caching for auto discovery type queries with result transformer

### LINKED ISSUE #3169: InvalidOperationException: This transformer is not initialized by Cached Query
Between version 5.3.10 and 5.3.11 the behaviour of cached SQL Query results changed when using AliasToBean transform. 

```csharp
ISession session;
session.CreateSQLQuery("select 'REGIONCODE' as regionCode  ")
		.AddScalar("regionCode", NHibernateUtil.String)
		.SetResultTransformer(Transformers.AliasToBean<ResultDto>())
		.SetCacheable(true)
		.List();
```

This works the first time through when caching is enabled, but when run again it fails with a "System.InvalidOperationException : This transformer is not initialized" exception.

I've created the attached Test Case which gives the exception in the second call to .List in the GetCacheableSqlQueryResults method.

In my actual code the SELECT statement involves tables and join, etc - its been simplified for this test case to not include any entities.  

[GH3169.zip](https://github.com/nhibernate/nhibernate-core/files/9657395/GH3169.zip)

### PR DESCRIPTION
**Pull request by bahusoid**
Fixes #3169  

Exception was triggered by changes in 5.3.10 (https://github.com/nhibernate/nhibernate-core/pull/2943)  (see details https://github.com/nhibernate/nhibernate-core/issues/3169#issuecomment-1262101953)

### CODE REVIEW

File: src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+						.List<ResultDto>();
+			}
+
+			using (var session = OpenSession())
+			{
+				using (EnableStatisticsScope())
+				{
+					var l = GetCacheableSqlQueryResults(session);
+					Assert.AreEqual(1, l.Count);
+					//Uncomment if we properly fix caching auto discovery type queries with transformers
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs:43]:
```suggestion
					//TODO: Uncomment if we properly fix caching auto discovery type queries with transformers
```

File: src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+		protected override HbmMapping GetMappings()
+		{
+			var mapper = new ModelMapper();
+			return mapper.CompileMappingForAllExplicitlyAddedEntities();
+		}
+
+
+		[Test]
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs:24]:
```suggestion
		}

		[Test]
```

File: src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+					//Uncomment if we properly fix caching auto discovery type queries with transformers
+					//Assert.That(Sfi.Statistics.QueryCacheMissCount, Is.EqualTo(1), "results are expected from DB");
+					//Assert.That(Sfi.Statistics.QueryCacheHitCount, Is.EqualTo(0), "results are expected from DB");
+				}
+
+				using (EnableStatisticsScope())
+				{
+					var l2 = GetCacheableSqlQueryResults(session);
+					Assert.AreEqual(1, l2.Count);
+					//Uncomment if we properly fix caching auto discovery type queries with transformers
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs:52]:
```suggestion
					//TODO: Uncomment if we properly fix caching auto discovery type queries with transformers
```

File: src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		}
+
+
+		[Test]
+		public void CachedQueryWithTransformer()
+		{
+			IList<ResultDto> GetCacheableSqlQueryResults(ISession s)
+			{
+				return s.CreateSQLQuery(
+							"select 'REGIONCODE' as regionCode  ")
```
Comment (hazzik) [inline src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs:30]: Certain DBs (mainly PLSQL based like Oracle, Postgres, Firebird) require a table name.

Comment (hazzik) [reply src/NHibernate.Test/NHSpecificTest/GH3169/FixtureByCode.cs:30]: https://stackoverflow.com/questions/3732422/select-from-nothing

**Comment by hazzik** [comment]
Can you move tests to NativeSQLQueriesFixture.cs as a copy of this test?
https://github.com/nhibernate/nhibernate-core/blob/ba574e949c6c382503d346f63d69fd9cadc2dcb2/src/NHibernate.Test/SqlTest/Query/NativeSQLQueriesFixture.cs#L208-L210

**Review by fredericDelaporte — APPROVED** [review]

File: src/NHibernate.Test/SqlTest/Query/NativeSQLQueriesFixture.cs [hunk 5]
```diff
@@ -264,6 +264,56 @@ IList GetCacheableSqlQueryResults()
...
+					Assert.AreEqual(1, l.Count);
+					//TODO: Uncomment if we properly fix caching auto discovery type queries with transformers
+					// var msg = "results are expected from " + (fromCache ? "cache" : "DB");
+					// Assert.That(Sfi.Statistics.QueryCacheMissCount, Is.EqualTo(fromCache ? 0 : 1), msg);
+					// Assert.That(Sfi.Statistics.QueryCacheHitCount, Is.EqualTo(fromCache ? 1 : 0), msg);
+				}
+			}
+
+
+			AssertQuery(false);
```
Comment (fredericDelaporte) [inline src/NHibernate.Test/SqlTest/Query/NativeSQLQueriesFixture.cs:306]:
```suggestion
			}

			AssertQuery(false);
```

**Comment by fredericDelaporte** [comment]
We may still support caching, see bahusoid/nhibernate-core#28.

**Comment by bahusoid** [comment]
> We may still support caching, see https://github.com/bahusoid/nhibernate-core/pull/28.

Not trivial fix. Let's do it in 5.4.

**Review by hazzik — APPROVED** [review]

**Pull request merged**